File: outside_code/exact_v0.21/code/C_version/Darwin.c
// Darwin.c
// Estimate p-value for Darwin's finches dataset, using Roberts & Stone (1990)'s $\bar S^2$ statistic.

// Got p-value = 0.000463 for n_samples = 10000000

// parameters
int n_samples = 100000;
int matrix_type = 0;

// observed matrix
#define N_ROWS 13
#define N_COLS 17
int observed_matrix[N_ROWS][N_COLS] = {
    {0,0,1,1,1,1,1,1,1,1,0,1,1,1,1,1,1},
    {1,1,1,1,1,1,1,1,1,1,0,1,0,1,1,0,0},
    {1,1,1,1,1,1,1,1,1,1,1,1,0,1,1,0,0},
    {0,0,1,1,1,0,0,1,0,1,0,1,1,0,1,1,1},
    {1,1,1,0,1,1,1,1,1,1,0,1,0,1,1,0,0},
    {0,0,0,0,0,0,0,0,0,0,1,0,1,0,0,0,0},
    {0,0,1,1,1,1,1,1,1,0,0,1,0,1,1,0,0},
    {0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0},
    {0,0,1,1,1,1,1,1,1,1,0,1,0,0,1,0,0},
    {0,0,1,1,1,1,1,1,1,1,0,1,0,1,1,0,0},
    {0,0,1,1,1,0,1,1,0,1,0,0,0,0,0,0,0},
    {0,0,1,1,0,0,0,0,0,0,0,0,0,0,0,0,0},
    {1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1}};

// Return an upper bound for the test statistic on a m-by-n matrix.
// The minimum possible value is assumed to be 0.
int max_statistic(int m, int n) {
    return n*n*(m*(m-1))/2;
}

// Compute the test statistic for a m-by-n matrix.
// (Must return an integer between 0 and max_statistic(m,n).)
int compute_statistic(int** matrix, int* p, int* q, int m, int n) {
    // This function computes nchoosek(m,2) times $\bar S^2$.
    int s = 0;
    for (int i = 0; i<m-1; i++) {
        for (int j = i+1; j<m; j++) {
            int C_ij = 0;
            for (int k = 0; k<n; k++) {
                C_ij += matrix[i][k]*matrix[j][k];
            }
            s += C_ij*C_ij;
        }
    }
    return s;
}


// Import main() for computing p-values
#include "p_value.c"

File: outside_code/exact_v0.21/examples/null models in ecology/Darwin.m
% Use the exact sampler to estimate the p-value for Darwin's finches dataset
% using Roberts & Stone (1990)'s $\bar S^2$ statistic.
fprintf('\n============== Null model analysis (Exact): Darwin''s finches ==============\n');

% number of samples to use
n_samples = 10000;
% confidence (for exact confidence interval)
confidence = 0.95;

% Darwin's finches
% 13 species (rows) and 17 islands (columns)
% Note: For the statistic below, we assume (rows,columns) correspond to (species,islands) respectively.
matrix_type = 0; % 0:binary, 1:nonnegative integer
A = [0 0 1 1 1 1 1 1 1 1 0 1 1 1 1 1 1
     1 1 1 1 1 1 1 1 1 1 0 1 0 1 1 0 0
     1 1 1 1 1 1 1 1 1 1 1 1 0 1 1 0 0
     0 0 1 1 1 0 0 1 0 1 0 1 1 0 1 1 1
     1 1 1 0 1 1 1 1 1 1 0 1 0 1 1 0 0
     0 0 0 0 0 0 0 0 0 0 1 0 1 0 0 0 0
     0 0 1 1 1 1 1 1 1 0 0 1 0 1 1 0 0
     0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0
     0 0 1 1 1 1 1 1 1 1 0 1 0 0 1 0 0
     0 0 1 1 1 1 1 1 1 1 0 1 0 1 1 0 0
     0 0 1 1 1 0 1 1 0 1 0 0 0 0 0 0 0
     0 0 1 1 0 0 0 0 0 0 0 0 0 0 0 0 0
     1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1];
p = sum(A,2)';
q = sum(A,1);
assert(all(p==[14 13 14 10 12 2 10 1 10 11 6 2 17]));
assert(all(q==[4 4 11 10 10 8 9 10 8 9 3 10 4 7 9 3 3]));

% Compute the test statistic on the input data
m = length(p); % number of species
C = A*A';
S = sum(sum(triu(C.^2)))/nchoosek(m,2);

% Generate lookup table for binary matrices with row sums p and column sums q
fprintf('Counting...\n');
count(p,q,matrix_type);

% Draw i.i.d. samples from the uniform distribution
fprintf('Sampling %d matrices...\n',n_samples);
X = sample(p,q,n_samples);

% Compute the test statistic on the samples
statistics = zeros(n_samples,1);
for i = 1:n_samples
    C = X(:,:,i)*X(:,:,i)';
    statistics(i) = sum(sum(triu(C.^2)))/nchoosek(m,2);
end

% Compute the p-value
p_value = mean(statistics >= S);
fprintf('Estimated p-value = %f\n',p_value);

% Compute an exact confidence interval for the p-value
[lower,upper] = BinomialCI(n_samples*p_value, n_samples, confidence, 1e-12);
fprintf('Exact %f%% confidence interval: (%f, %f)\n',100*confidence, lower,upper);
fprintf('(%f%% of the time, this interval will contain the true p-value.)\n',100*confidence);











